contracts/access/mod.go

```diff
@@ -0,0 +1,171 @@
+// Package access implements a native contract to handle access. It allows an
+// authorized identity to add access as an {ID, CONTRACT, COMMAND, IDENTITIES}
+// quadruplet.
+//
+// ID is the credential identifier. This identifier is generally defined at the
+// contract's creation.
+// CONTRACT is the contract name.
+// COMMAND specifies the command to grant access to on the contract.
+// IDENTITIES is a list of base64 encoded bls public keys, separated by comas.
+package access
+
+import (
+	"encoding/base64"
+	"encoding/hex"
+	"strings"
+
+	"go.dedis.ch/dela"
+	"go.dedis.ch/dela/core/access"
+	"go.dedis.ch/dela/core/execution"
+	"go.dedis.ch/dela/core/execution/native"
+	"go.dedis.ch/dela/core/store"
+	"go.dedis.ch/dela/crypto/bls"
+	"golang.org/x/xerrors"
+)
+
+const (
+	// ContractName is the name of the access contract.
+	ContractName = "go.dedis.ch/dela.Access"
+
+	// GrantIDArg is the argument's name in the transaction that contains the
+	// provided id to grant
+	GrantIDArg = "access:grant_id"
+
+	// GrantContractArg is the argument's name in the transaction that contain
+	// the provided contract name to grant the access to.
+	GrantContractArg = "access:grant_contract"
+
+	// GrantCommandArg is the argument's name in the transaction that contains
+	// the provided command to grant access to.
+	GrantCommandArg = "access:grant_command"
+
+	// IdentityArg is the argument's name in the transaction that contains the
+	// provided identity to grant access to.
+	IdentityArg = "access:identity"
+
+	// CmdArg is the argument's name to indicate the kind of command we want to
+	// run on the contract. Should be one of the Command type.
+	CmdArg = "access:command"
+)
+
+// Command defines a command for the value contract
+type Command string
+
+const (
+	// CmdSet defines the command to grant access
+	CmdSet Command = "GRANT"
+)
+
+// NewCreds creates new credentials for a value contract execution.
+func NewCreds(id []byte) access.Credential {
+	return access.NewContractCreds(id, ContractName, "all")
+}
+
+// RegisterContract registers the access contract to the given execution
+// service.
+func RegisterContract(exec *native.Service, c Contract) {
+	exec.Set(ContractName, c)
+}
+
+// Contract is the access contract that allows one to handle access.
+//
+// - implements native.Contract
+type Contract struct {
+	// access is the access service that will be modified.
+	access access.Service
+
+	// accessKey is the credential's ID allowed to use this smart contract
+	accessKey []byte
+
+	store store.Readable
+}
+
+// NewContract creates a new Value contract
+func NewContract(aKey []byte, srvc access.Service, store store.Readable) Contract {
+	return Contract{
+		access:    srvc,
+		accessKey: aKey,
+		store:     store,
+	}
+}
+
+// Execute implements native.Contract
+func (c Contract) Execute(snap store.Snapshot, step execution.Step) error {
+	creds := NewCreds(c.accessKey)
+
+	err := c.access.Match(c.store, creds, step.Current.GetIdentity())
```

Comment:
Continuing from above, is this check for ensuring the transaction executor's identity is allowed to execute a command in the access contract?

I'm not quite clear on where the initial permissions for the access contract are defined

Reply:
In `NewContract` we provide the `credential.Access` service, which is the one we use there. In the case of this contract, this service is not the regular access service that the blockchain uses, but a local one to the node, which is stored using `jstore.go`. You can see it in `controller/mod.go:63` when we create the contract. A "regular" contract would use to injector to retrieve the access service, but there we are creating a new local one, which can be updated locally with the cli. 

And yes, this check is to ensure that the identity has been (locally in this case) allowed to updated the access contract.

Reply:
> You can see it in `controller/mod.go:63` when we create the contract. A "regular" contract would use to injector to retrieve the access service, but there we are creating a new local one, which can be updated locally with the cli.

You probably meant resolving the access store there. So if I understand correctly, the invocation of the contract would result in all the local `jstores` being updated since the contract is executed and successful state changes are applied at all the nodes?

Btw, I couldn't an implementation of access.Service (probably `darc`?) being injected anywhere but the command assumes it's available to the injector. Am I overlooking something?

Reply:
In Dela there are now two different access services:

- the DARC one, which is the one used to control access to all smart contract, except the "access smart contract". The DARC access is the same on every node. It is kept in sync because we update it with transactions that we perform on the "access smart contract". The "access smart contract" is responsible for updating the DARC access. The DARC access is created [there](https://github.com/dedis/dela/blob/e1cc707548c7ab4d7cea0782ba76d2f24d8e2079/core/ordering/cosipbft/controller/mod.go#L122). It is injected in PR #174.

- the "local access", which is only used to control access on the "access smart contract". This access is local to every node and can be changed by running a CLI command locally. This access is stored in a json file (jstore.go).

In summary: the access smart contract is responsible for **updating the DARC access**. Access on the access smart contract is **not controlled by the DARC**, as it is the case for all other smart contracts, but by the **local access**, which we can update only with a local CLI command.

Reply:
That clears it! Thanks 👍